pwndbg/commands/hijack_fd.py

```diff
@@ -0,0 +1,283 @@
+
+    selected_protocol: Literal["tcp", "udp"] = "tcp"
+    selected_ip_protocol: Literal["ipv4", "ipv6"] | None = None
+    if parsed.scheme:
+        for any_value in scheme_info:
+            if any_value in ("tcp", "udp"):
+                selected_protocol = any_value
+            elif any_value in ("ipv4", "ipv6"):
+                selected_ip_protocol = any_value
+
+    domain_or_ip = parsed.hostname
+    if not domain_or_ip:
+        raise argparse.ArgumentTypeError("Domain or IP is required")
+
+    port = parsed.port
+    if not port:
+        raise argparse.ArgumentTypeError("Port is required")
+
+    protocol_ordered = [
+        ("ipv4", socket.AF_INET),
+        ("ipv6", socket.AF_INET6),
+    ]
+
+    found_ip_protocol: Literal["ipv4", "ipv6"] | None = None
+    address_ipv4_or_ipv6: str = ""
+    for family_name, family_const in protocol_ordered:
+        if selected_ip_protocol and selected_ip_protocol != family_name:
+            continue
+
+        try:
+            ips = socket.getaddrinfo(domain_or_ip, None, family_const)
+        except socket.gaierror:
+            # happen when domain not found
+            continue
+
+        for _, _, _, _, ip in ips:
+            address_ipv4_or_ipv6 = ip[0]
+            found_ip_protocol = family_name
+
+        if found_ip_protocol:
+            break
+
+    if not address_ipv4_or_ipv6:
+        raise argparse.ArgumentTypeError(
+            f"Could not resolve {domain_or_ip} to proper {selected_ip_protocol} address"
+        )
+
+    if not found_ip_protocol:
+        raise argparse.ArgumentTypeError("Protocol only accept: ipv4,ipv6")
+
+    return ParsedSocket(selected_protocol, found_ip_protocol, address_ipv4_or_ipv6, port)
+
+
+PARSED_FILE_ARG = Tuple[Optional[ParsedSocket], Optional[str]]
+
+
+def parse_file_or_socket(s: str) -> PARSED_FILE_ARG:
+    # is file
+    if s.startswith("/"):
+        return None, s
+    return parse_socket(s), None
+
+
+parser.add_argument(
+    "newfile",
+    help="""Specify a file or a socket.
+
+For files, the filename must start with `/` (e.g., `/etc/passwd`).
+
+For sockets, the following formats are allowed:
+- `127.0.0.1:80` (default is TCP)
+- `tcp://[::1]:80`
+- `udp://example.com:80`
+- `tcp+ipv6://example.com:80`
+    """,
+    type=parse_file_or_socket,
+)
+
+
+@pwndbg.commands.ArgparsedCommand(parser, category=CommandCategory.MISC, command_name="hijack-fd")
+@pwndbg.commands.OnlyWhenRunning
+@pwndbg.commands.OnlyWhenUserspace
+def hijack_fd(fdnum: int, newfile: PARSED_FILE_ARG) -> None:
+    socket_data, filename = newfile
+    if filename:
+        stack_size, asm_bin = asm_replace_file(fdnum, filename)
+    elif socket_data:
+        stack_size, asm_bin = asm_replace_socket(fdnum, socket_data)
+    else:
+        assert False
+
+    async def ctrl(ec: pwndbg.dbg_mod.ExecutionController):
+        async with exec_shellcode_with_stack(ec, asm_bin, stack_size):
+            print("Operation probably succeeded. Errors are not captured.")
```

Comment:
Probably? Why probably?:)

Reply:
we don't capture errors inside shell code, when syscall fail